machineconfiguration/v1alpha1/types_machineosconfig.go

```diff
@@ -0,0 +1,233 @@
+	// this push secret will be used only by the MachineConfigController pod to push the image to the final destination. Not all nodes will need to push this image, most of them
+	// will only need to pull the image in order to use it.
+	// +kubebuilder:validation:Required
+	RenderedImagePushSecret ImageSecretObjectReference `json:"renderedImagePushSecret"`
+	// renderedImagePullSecret is the secret used to pull the final produced image.
+	// the final image push and pull secrets should be separate for security concerns. If the final image push secret is somehow exfiltrated,
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this pull secret will be used on all nodes in the pool. These nodes will need to pull the final OS image and boot into it using rpm-ostree or bootc.
+	// +optional
+	RenderedImagePullSecret ImageSecretObjectReference `json:"renderedImagePullSecret"`
+	// renderedImagePullspec describes the location of the final image.
+	// the MachineOSConfig object will use the in cluster image registry configuration.
+	// if you wish to use a mirror or any other settings specific to registries.conf, please specify those in the cluster wide registries.conf.
+	// The format of the image pullspec is:
+	// host[:port][/namespace]/name@sha256:<digest> or host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`((self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$')) || (self.split(':').size() == 2 && self.split(':')[1].matches('^([a-zA-Z0-9-./:])+$')) || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix or a valid :<tag>, where '<digest>' is 64 characters long and '<tag>' is any valid string  Or it must be a valid .svc followed by a port, repository, image name, and tag."
+	// +kubebuilder:validation:XValidation:rule=`((self.split('@').size() == 2 && self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$')) || (self.split(':').size() == 2 && self.split(':')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$')) || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme. Or it must be a valid .svc followed by a port, repository, image name, and tag."
+	// +kubebuilder:validation:Required
+	RenderedImagePullspec string `json:"renderedImagePullspec"`
```

Comment:
from the perspective of the machineosconfig, this is an output, not an input.  it cannot SHA based since we don't know the sha in advance. And the spec is strictly a push spec, not a pull spec.

Reply:
update to require being a tag.  Update to indicate it's a push spec.